File: tests/test_race_contracts.js
'use strict';
const assert=require('assert');
const fs=require('fs');
const vm=require('vm');
const contracts=require('../docs/assets/race-contracts.js');
const replay=require('../docs/assets/runner-replay.js');
const media=require('../docs/assets/race-media.js');
const audience=require('../docs/assets/audience-analytics.js');
const map=require('../docs/assets/map.js');
const registry=require('../data/routes/ultravasan90-routes.json');

for(const [key,edition] of Object.entries(contracts.catalog.editions)){
  const misleading={race_key:key,name:'45 90 another event',year:1900,distance_km:1,race_family:'wrong',course_version:'wrong'};
  assert.strictEqual(contracts.familyForRace(misleading),edition.race_family);
  assert.strictEqual(audience.audienceRaceFamily(misleading),edition.race_family);
  assert.strictEqual(map.mapRaceFamily(misleading),edition.race_family);
  assert.strictEqual(media.familyForRace(misleading),edition.race_family);
  assert.strictEqual(contracts.supports(misleading,'replay'),edition.capabilities.replay);
  assert.strictEqual(contracts.supports(misleading,'map_duel'),edition.capabilities.map_duel);
  assert.strictEqual(contracts.supports(misleading,'not-a-capability'),false);
  assert.strictEqual(replay.routeForRace(registry,misleading),registry.routes[contracts.courseForRace(key).display_route_id]);
  assert.strictEqual(replay.medalTimeForRace(misleading,'M'),edition.medal_profile==='pre2023'?34199:edition.medal_profile==='post2023'?35999:null);
}
for(const key of ['ultravasan90-2099','ultravasan45-2099','uv90','constructor','__proto__','']){
  const unknown={race_key:key,name:'Ultravasan 90',year:2025,distance_km:90};
  assert.strictEqual(contracts.familyForRace(unknown),null);
  assert.strictEqual(replay.routeForRace(registry,unknown),null);
  assert.strictEqual(replay.medalTimeForRace(unknown,'M'),null);
  assert.strictEqual(media.musicForRace(unknown),null);
  assert.throws(()=>contracts.assertKnownRaces([unknown]),/saknar kontrakt/);
}
assert.strictEqual(media.musicForRace('uv45'),contracts.family('uv45').music,'Explicit family-level media selection remains supported');
assert.ok(map.mixedRaceFamilyError([{race_id:999}],[]),'Missing race contract must stop a map duel');
assert.strictEqual(map.activeReferenceRoute([],[],registry),null,'No default UV90 route for missing selection');

// Another event and opaque edition key work without changing the resolver.
const other={schema_version:1,event:{event_key:'other'},families:{short:{event_key:'other'}},courses:{v7:{event_key:'other',race_family:'short',display_route_id:'shape-b'}},editions:{'opaque/id':{race_key:'opaque/id',event_key:'other',race_family:'short',course_version_id:'v7',medal_profile:null,capabilities:{replay:false,map_duel:false}}}};
const custom=contracts.create(other),shape={id:'shape-b'};
assert.strictEqual(custom.familyForRace({race_key:'opaque/id',name:'Ultravasan 90',year:2037,distance_km:90}),'short');
assert.strictEqual(custom.routeForRace({routes:{'shape-b':shape}},'opaque/id'),shape);
assert.strictEqual(custom.supports('opaque/id','replay'),false);
other.editions['opaque/id'].race_family='uv90';
assert.strictEqual(custom.familyForRace('opaque/id'),'short','Input mutation cannot change snapshot');
assert.throws(()=>{custom.catalog.editions['opaque/id'].race_family='uv90'},TypeError);
assert.throws(()=>contracts.create(other),/Ogiltigt loppkontrakt/);

// Same resolver and generated catalog also work through the browser globals.
const browser={window:{}};
for(const file of ['../docs/data/race-catalog.js','../docs/assets/race-contracts.js','../docs/assets/runner-replay.js','../docs/assets/race-media.js'])vm.runInNewContext(fs.readFileSync(require.resolve(file),'utf8'),browser);
assert.strictEqual(browser.window.RunnerReplay.routeForRace(registry,{race_key:'ultravasan45-2014',year:2099}).id,'ultravasan45-current');
assert.strictEqual(browser.window.RACE_MEDIA_CONFIG.familyForRace({race_key:'ultravasan90-2022',name:'45'}),'uv90');
assert.strictEqual(browser.window.RaceContracts.courseForRace('ultravasan90-2026').checkpoint_catalog.find(cp=>cp.checkpoint_key==='high_point').distance_km,null);
console.log('OK: explicit contracts across all 22 editions, unknown keys, immutable snapshots and another event');

File: tests/test_race_contracts.py
"""Contract boundaries: raw-data compatibility, immutability and explicit routing."""
from copy import deepcopy

import pytest

from tools import race_contracts as contracts
from tools.uvtool import validation_rule_for_race


@pytest.fixture
def inputs():
    root = contracts.ROOT
    return dict(config=contracts.read_json(root / "config/races.json"),
                definitions=contracts.read_json(root / "config/course_versions.json"),
                lock=contracts.read_json(root / contracts.LOCK_PATH),
                registry=contracts.read_json(root / "data/routes/ultravasan90-routes.json"),
                observed=contracts.observed_editions())


def test_all_observed_editions_resolve_without_changing_raw_data(inputs):
    contracts.verify_route_export(inputs["registry"])
    catalog = contracts.build_catalog(**inputs)
    assert len(catalog["editions"]) == 22
    assert len(catalog["courses"]) == 5
    assert {e["race_family"] for e in catalog["editions"].values()} == {"uv90", "uv45"}
    for path, expected in contracts.serialized_catalog(catalog).items():
        assert (contracts.ROOT / path).read_text(encoding="utf-8") == expected
    old = catalog["editions"]["ultravasan90-2025"]["course_version_id"]
    new = catalog["editions"]["ultravasan90-2026"]["course_version_id"]
    assert old != new
    assert catalog["courses"][old]["display_route_id"] == catalog["courses"][new]["display_route_id"]
    course = catalog["courses"][new]
    assert {cp["checkpoint_key"] for cp in course["checkpoint_catalog"] if cp["distance_km"] is None} == {"high_point", "mora_warning"}
    assert [segment["distance_km"] for segment in course["segments"] if segment["from"] == "high_point"] == [None]
    legacy = catalog["courses"]["uv45-2014-2024-v1"]
    assert legacy["geometry_source"]["usage"] == "display-reference"
    assert legacy["whole_course_comparison_group"] is None
    assert legacy["checkpoint_catalog"][1]["distance_km"] == 15.5


def test_whole_course_group_fails_closed_until_edition_equivalence_is_verified(inputs):
    catalog = contracts.build_catalog(**inputs)
    for year in (2023, 2024, 2025, 2026):
        edition = catalog["editions"][f"ultravasan90-{year}"]
        assert edition["whole_course_comparison_group"] is None
    assert catalog["editions"]["ultravasan90-2022"]["whole_course_comparison_group"] is None
    assert catalog["editions"]["ultravasan45-2026"]["whole_course_comparison_group"] is None
    assert all(course["whole_course_comparison_group"] is None for course in catalog["courses"].values()), (
        "a checkpoint CourseVersion must not own a whole-course comparison decision"
    )


@pytest.mark.parametrize("change", ["missing_edition", "wrong_family", "wrong_event", "unknown_course", "duplicate", "observed_control"])
def test_invalid_assignments_fail_closed(inputs, change):
    race = inputs["config"]["races"][0]
    if change == "missing_edition":
        inputs["config"]["races"].pop(0)
    elif change == "duplicate":
        inputs["config"]["races"].append(deepcopy(race))
    elif change == "wrong_family":
        race["race_family"] = "uv45" if race["race_family"] == "uv90" else "uv90"
    elif change == "wrong_event":
        race["event_key"] = "different-event"
    elif change == "unknown_course":
        race["course_version_id"] = "missing"
    else:
        inputs["observed"][0]["checkpoints"][1]["distance_km"] += 0.1
    with pytest.raises(ValueError):
        contracts.build_catalog(**inputs)


@pytest.mark.parametrize("change", ["checkpoint", "anchor", "segment", "display_geometry", "family", "source"])
def test_material_mutations_cannot_reuse_course_id(inputs, change):
    key = "uv90-pre2023-v1"
    definition = inputs["definitions"]["courses"][key]
    route = inputs["registry"]["routes"][definition["display_route_id"]]
    if change == "checkpoint":
        definition["checkpoint_catalog"][1]["distance_km"] += .01
        definition["segments"] = contracts.expected_segments(definition["checkpoint_catalog"])
    elif change == "anchor":
        definition["display_anchors"][1]["coord"][0] += .001
        route["checkpoints"] = deepcopy(definition["display_anchors"])
    elif change == "segment":
        definition["segments"][0]["to"] = "eldris"
    elif change == "display_geometry":
        route["points"][1][0] += .001
    elif change == "family":
        definition["race_family"] = "uv45"
    else:
        definition["geometry_source"]["usage"] = "verified-historical-geometry"
    with pytest.raises(ValueError):
        contracts.build_catalog(**inputs)


def test_rehashing_existing_id_cannot_bypass_published_lock(inputs):
    key = "uv90-pre2023-v1"
    previous = deepcopy(inputs["lock"]["courses"])
    definition = inputs["definitions"]["courses"][key]
    definition["geometry_source"]["note"] = "changed material provenance"
    inputs["lock"]["courses"][key] = contracts.digest(contracts.course_material(key, definition, inputs["registry"]))
    with pytest.raises(ValueError, match="create a new ID"):
        contracts.verify_append_only(previous, inputs["lock"]["courses"])
    contracts.verify_append_only(previous, {**previous, "new-version-v2": "new-fingerprint"})
    with pytest.raises(ValueError, match="removed"):
        contracts.verify_append_only(previous, {})


def test_gpx_fingerprint_ignores_timestamps_but_keeps_points_elevation_and_segments(tmp_path):
    path = tmp_path / "geometry.gpx"
    source = '<gpx xmlns="http://www.topografix.com/GPX/1/1"><metadata><time>2024</time></metadata><trk><trkseg><trkpt lat="61" lon="13"><ele>20</ele><time>one</time></trkpt><trkpt lat="62" lon="14"><ele>30</ele></trkpt></trkseg></trk></gpx>'
    path.write_text(source)
    original = contracts.gpx_geometry_digest(path)
    path.write_text(source.replace("2024", "2026").replace("one", "two").replace("><", ">\n<").replace('lat="61"', 'lat="61.0000"'))
    assert contracts.gpx_geometry_digest(path) == original
    for mutation in [source.replace('lat="61"', 'lat="61.001"'), source.replace('<ele>20</ele>', '<ele>21</ele>'), source.replace('</trkpt><trkpt', '</trkpt></trkseg><trkseg><trkpt')]:
        path.write_text(mutation)
        assert contracts.gpx_geometry_digest(path) != original


def test_validation_uses_explicit_family_and_never_prefix(inputs):
    config = inputs["config"]
    assert validation_rule_for_race(config, {"race_key": "ultravasan45-2025"})["finish_seconds_min"] == 5400
    assert validation_rule_for_race(config, {"race_key": "ultravasan45-2099"}) == {}
    config["races"].append({"race_key": "opaque-key", "race_family": "uv90"})
    assert validation_rule_for_race(config, {"race_key": "opaque-key", "name": "45", "distance_km": 45})["finish_seconds_min"] == 14400


def test_browser_must_receive_the_fingerprinted_geometry(inputs):
    inputs["registry"]["routes"]["ultravasan90-pre2023"]["points"][1][0] += .001
    with pytest.raises(ValueError, match="Browser routes differ"):
        contracts.verify_route_export(inputs["registry"])
